cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+under "programmable matching on virtual
+nodes"](https://github.com/bytecodealliance/rfcs/blob/main/accepted/cranelift-isel-isle-peepmatic.md#extractors-programmable-matching-on-virtual-nodes).
+
+To provide a concrete example, if we have the term declarations
+
+```lisp
+    (decl A (u32 u32) T)
+    (decl B (T) U)
+```
+
+then if we write a rule like
+
+```lisp
+    (rule (B (A x y))
+          (U.Variant1 x y))
+```
+
+then we have used `A` as an *extractor*. When `B` is invoked as a
+constructor with some `T`, this rule uses `A` as an extractor and
+attempts (via whatever programmable matching behavior) to use `A` to
+turn the `T` into two `u32`s, binding `x` and `y`. `A` can succeed or
+fail, just as any other part of a pattern-match can.
+
+Just as for constructors, there are *internal* and *external*
+extractors. Most of the interesting programmable behavior occurs in
+external extractors, which are defined in Rust; we will discuss this
+further in a section below. Internal extractors, in contrast, behave
+like macros, and can be defined for convenience: for example, we can
+write
+
+```lisp
+    (decl A (u32 u32) T)
+    (extractor (A pat1 pat2)
+               (and
+                 (extractArg1 pat1)
+                 (extractArg2 pat2)))
+```
+
+which will, for example, expand a pattern `(A (subterm ...) _)` into
+`(and (extractArg1 (subterm ...)) (extractArg2 _))`.
+
+#### Summary: Terms, Constructors, and Extractors
+
+We start with a `term`, which is just a schema for data:
+
+```lisp
+    (decl Term (A B C u32 u32) T)
+```
+
+A term can have:
+
+1. A single internal extractor body, via a toplevel `(extractor ...)`
+   form, OR
+   
+2. A single external extractor binding, via an `(extern extractor
+   ...)` form, defined below; AND
+   
+3. One or more `(rule (Term ...) ...)` toplevel forms, which together
+   make up an internal constructor definition, OR
+   
+4. A single external constructor binding, via an `(extern constructor
+   ...)` form, defined below.
+
+## ISLE to Rust
+
+Now that we have described the core ISLE language, we will document
+how it interacts with Rust code. We consider these interactions to be
+semantically as important as the core language: they are not
+implementation details, but rather, a well-defined interface by which
+ISLE can interface with the outside world (an "FFI" of sorts).
+
+### Mapping to Rust: Constructors, Functions, and Control Flow
+
+ISLE was designed to have a simple, easy-to-understand mapping from
+its language semantics to Rust semantics. This means that the
+execution of ISLE rewriting has a well-defined implementation in
+Rust. The basic principles are:
+
+1. Every constructor with rules in ISLE becomes a Rust function. The
+   constructor arguments are the Rust function arguments. The
+   constructor's "return value" is the Rust function's return value
+   (usually wrapped in an `Option` because pattern coverage can be
```

Comment:
Why usually and not always?

Reply:
My mistake, it actually is always, in this context; I was thinking of constructor signatures in general, as they are infallible when imported (Rust code called from the RHS is not allowed to fail/backtrack). In this context we're only discussing internal constructors, so that doesn't apply; updated.